### PR TITLE
GraphQL Subscriptions

### PR DESCRIPTION
**Pull request by dushaniw**
Fix wso2/product-apim#11574

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * This method process websocket handshake and perform authentication using the inbound message context.
+     * For successful authentications, it sets the resource map of the invoking API to the context.
+     *
+     * @param inboundMessageContext InboundMessageContext
+     * @return InboundProcessorResponseDTO with handshake processing response
+     */
+    public InboundProcessorResponseDTO processHandshake(InboundMessageContext inboundMessageContext) {
+
+        if(log.isDebugEnabled()) {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java:52]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                    WebSocketApiConstants.HandshakeErrorConstants.API_AUTH_ERROR,
+                    WebSocketApiConstants.HandshakeErrorConstants.API_AUTH_GENERAL_MESSAGE);
+        } catch (APISecurityException e) {
+            log.error(e);
+            return InboundWebsocketProcessorUtil.getHandshakeErrorDTO(
+                    WebSocketApiConstants.HandshakeErrorConstants.API_AUTH_ERROR,
+                    e.getMessage());
+        }
+        if (isOAuthHeaderValid) {
+            if(log.isDebugEnabled()) {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java:72]: Formatting issue

Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java (hunk 1 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/handshake/HandshakeProcessor.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+     * @throws APIManagementException if an internal error occurs
+     * @throws APISecurityException   if authentication fails
+     */
+    public static boolean authenticateWSJWTToken(InboundMessageContext inboundMessageContext)
+            throws APIManagementException, APISecurityException {
+
+        AuthenticationContext authenticationContext;
+        JWTValidator jwtValidator = new JWTValidator(new APIKeyValidator(), inboundMessageContext.getTenantDomain());
+        authenticationContext = jwtValidator.
+                authenticateForWebSocket(inboundMessageContext.getSignedJWTInfo(),
```
Comment (wasuradananjith) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:154]: Can't we only pass inboundMessageContext to this function, so we can get the already set parameters inside the function itself?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+    public static boolean authenticateGraphQLJWTToken(InboundMessageContext inboundMessageContext)
+            throws APIManagementException, APISecurityException {
+
+        AuthenticationContext authenticationContext;
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(inboundMessageContext.getTenantDomain(),
+                true);
+        JWTValidator jwtValidator = new JWTValidator(new APIKeyValidator(), inboundMessageContext.getTenantDomain());
+        authenticationContext = jwtValidator.
+                authenticateForGraphQLSubscription(inboundMessageContext.getSignedJWTInfo(),
```
Comment (wasuradananjith) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:135]: Here as well, Can't we pass only inboundMessageContext?

Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:135]: IMO JWTValidator logic should not depend on the transport specific message context InboundMessageContext or SynapseMessageContext, hence passing the signed JWT directly is better.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java (hunk 3 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:154]: same answer as below

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java (hunk 4 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:135]: +1

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java (hunk 3 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/inbound/websocket/utils/InboundWebsocketProcessorUtil.java:154]: +1

File: components/apimgt/org.wso2.carbon.apimgt.api/pom.xml [hunk 5]
```diff
@@ -43,6 +43,11 @@
             <version>4.5.2</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.graphQL</groupId>
+            <artifactId>graphQL</artifactId>
+            <version>${graphql.java.version}</version>
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.api/pom.xml:49]: remove version here

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml [hunk 6]
```diff
@@ -53,6 +53,21 @@
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.graphQL</groupId>
+            <artifactId>graphQL</artifactId>
+            <version>${graphql.java.version}</version>
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml:59]: remove version and move to dependency management

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml [hunk 7]
```diff
@@ -53,6 +53,21 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.googlecode.json-simple.wso2</groupId>
+            <artifactId>json-simple</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>commons-io.wso2</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>${commons-io.version}</version>
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml:68]: same goes here

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/QueryAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+import graphql.GraphQL;
+import graphql.GraphQLError;
+import graphql.analysis.FieldComplexityCalculator;
+import graphql.analysis.MaxQueryComplexityInstrumentation;
+import graphql.analysis.MaxQueryDepthInstrumentation;
+import graphql.schema.GraphQLSchema;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.apimgt.common.gateway.dto.QueryAnalyzerResponseDTO;
+
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/QueryAnalyzer.java:31]: remove newline

File: components/apimgt/org.wso2.carbon.apimgt.api/pom.xml (hunk 5 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.api/pom.xml:49]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml (hunk 6 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml:59]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml (hunk 7 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.common.gateway/pom.xml:68]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/QueryAnalyzer.java (hunk 8 above)
Comment (dushaniw) [reply components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/QueryAnalyzer.java:31]: fixed

**Review by tharindu1st — APPROVED** [review]

**Review by 1akshitha — APPROVED** [review]

**Pull request merged**